module/bdev/aio/bdev_aio_sync.c

```diff
@@ -0,0 +1,179 @@
+#include "bdev_aio.h"
+#include "spdk/env.h"
+#include "spdk/stdinc.h"
+#include "spdk/thread.h"
+#include "spdk/bdev_module.h"
+#include "spdk/log.h"
+
+#include <libaio.h>
+
+#include "bdev_aio_task.h"
+#include "bdev_aio_sync.h"
+
+#define MAX_QUEUE_LEN 1024
+
+static struct spdk_ring *g_blocking_ring = NULL;
+static pthread_t g_blocking_worker_thread;
+static pthread_mutex_t g_mutex;
+static pthread_cond_t g_cond;
+static bool g_exit;
+
+static void
+aio_complete(struct spdk_bdev_io *bdev_io, int status, int aio_errno)
+{
+	if (status == 0) {
+		spdk_bdev_io_complete(bdev_io, SPDK_BDEV_IO_STATUS_SUCCESS);
+	} else {
+		spdk_bdev_io_complete_aio_status(bdev_io, -aio_errno);
+	}
+}
+
+static void
+aio_call_complete_fn(void *arg)
+{
+	struct aio_request_ctx *request = arg;
+
+	aio_complete(spdk_bdev_io_from_ctx(request->aio_task), request->status, request->aio_errno);
+
+	free(arg);
+}
+
+static void
+aio_call_request_fn(void *arg)
+{
+	struct aio_request_ctx *request = arg;
+
+	request->status = request->fn(request->ctx);
+	request->aio_errno = errno;
+
+	spdk_thread_send_msg(request->thread, aio_call_complete_fn, request);
+}
+
+#define BATCH_SIZE 64
+
+static void *
+aio_blocking_worker(void *arg)
+{
+	void *message[BATCH_SIZE];
+	size_t count;
+	size_t i;
+
+	SPDK_NOTICELOG("aio_blocking_worker started on core:%d\n", sched_getcpu());
+
+	pthread_mutex_lock(&g_mutex);
+
+	for (;;) {
+		for (;;) {
+			count = spdk_ring_dequeue(g_blocking_ring, message, BATCH_SIZE);
```

Comment:
As mentioned by @gila in https://github.com/openebs/spdk/pull/10#discussion_r564917506, these are lockless... You should not need the mutex.

Reply:
A mutex is necessary in order to call pthread_cond_wait(). However there is still a race condition that can occur if the queue is empty, and an element is inserted into the queue after the call to spdk_ring_dequeue() but before the pthread_cond_wait(). These two calls must therefore be made within a critical section, as must also the pthread_cond_signal().

This *is* a trade-off, to avoid having the worker thread continually busy when the queue is likely to be empty most of the time. No mutex is required to add things to the queue except in the one case where the queue was empty beforehand. While the dequeue does need to be in a critical section, it is batched and so the mutex only needs to be acquired once per batch.